packages/upload-client/src/receipts.js

```
@@ -0,0 +1,110 @@
+import retry, { AbortError } from 'p-retry'
+import { CAR } from '@ucanto/transport'
+import { receiptsEndpoint } from './service.js'
+import { REQUEST_RETRIES } from './constants.js'
+
+export class ReceiptNotFound extends Error {
+  /**
+   * @param {import('multiformats').UnknownLink} taskCid
+   */
+  constructor(taskCid) {
+    super()
+    this.taskCid = taskCid
+  }
+
+  /* c8 ignore start */
+  get reason() {
+    return `receipt not found for task ${this.taskCid}`
+  }
+  /* c8 ignore end */
+
+  get name() {
+    return 'ReceiptNotFound'
+  }
+}
+
+export class Receipt {
+  /**
+   * @param {import('./types.js').RequestOptions} [options]
+   */
+  constructor(options = {}) {
+    /* c8 ignore start */
+    this.receiptsEndpoint = options.receiptsEndpoint ?? receiptsEndpoint
+    this.retries = options.retries ?? REQUEST_RETRIES
+    this.fetch = options.fetch ?? globalThis.fetch.bind(globalThis)
+    /* c8 ignore stop */
+  }
+
+  /**
+   * Polls for a receipt for an executed task by its CID.
+   *
+   * @param {import('multiformats').UnknownLink} taskCid
+   * @returns {Promise<import('@ucanto/interface').Receipt>}
+   */
+  async poll(taskCid) {
+    return await retry(
+      async () => {
+        const res = await this.get(taskCid)
+        if (res.error) {
+          // @ts-ignore
+          if (res.error.name === 'ReceiptNotFound') {
+            // throw an error that will cause `p-retry` to retry with
+            throw new Error('blob/accept receipt not yet available')
```

can we propagate this error so as receipt not found so that consumer can know the real error and give different info to user? Like a not found would probably mean more of a problem, while other error may mean some downtime on the service

---

addressed